File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Util/Vector2d.java
package org.firstinspires.ftc.teamcode.Util;

public class Vector2d {
  public double x, y;

  public Vector2d(double x, double y) {
    this.x = x;
    this.y = y;
  }

  // =============================================

  public double norm() {
    return Math.sqrt(x * x + y * y);
  }

  public double angle() {
    return Math.atan2(y, x);
  }

  public Vector2d plus(Vector2d other) {
    return new Vector2d(x + other.x, y + other.y);
  }

  public Vector2d minus(Vector2d other) {
    return new Vector2d(x - other.x, y - other.y);
  }

  public Vector2d times(double scalar) {
    return new Vector2d(x * scalar, y * scalar);
  }

  public Vector2d div(double scalar) {
    return new Vector2d(x / scalar, y / scalar);
  }

  public Vector2d unaryMinus() {
    return new Vector2d(-x, -y);
  }

  public double dot(Vector2d other) {
    return x * other.x + y * other.y;
  }

  public double distanceTo(Vector2d other) {
    return this.minus(other).norm();
  }

  public Vector2d rotated(double angle) {
    // 10 * 1              - 0
    double newX = x * Math.cos(angle) - y * Math.sin(angle);
    // 10 * 0              + 0
    double newY = x * Math.sin(angle) + y * Math.cos(angle);
    return new Vector2d(newX, newY);
  }

  @Override
  public String toString() {
    return String.format("(%.3f, %.3f)", x, y);
  }

  public boolean equals(Vector2d other) {
    if (other instanceof Vector2d) {
      return Math.abs(x - other.x) < 1e-4 && Math.abs(y - other.y) < 1e-4;
    } else {
      return false;
    }
  }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Vector2d.kt
package org.firstinspires.ftc.teamcode.util

import kotlin.math.pow
import kotlin.math.sqrt

class Vector2d(var x: Double, var y: Double) {

    // =============================================

    fun norm(): Double {
        return Math.sqrt(x * x + y * y)
    }

    fun angle(): Double {
        return Math.atan2(y, x)
    }

    operator fun plus(other: Vector2d): Vector2d {
        return Vector2d(x + other.x, y + other.y)
    }

    operator fun minus(other: Vector2d): Vector2d {
        return Vector2d(x - other.x, y - other.y)
    }

    operator fun times(scalar: Double): Vector2d {
        return Vector2d(x * scalar, y * scalar)
    }

    operator fun div(scalar: Double): Vector2d {
        return Vector2d(x / scalar, y / scalar)
    }

    operator fun unaryMinus(): Vector2d {
        return Vector2d(-x, -y)
    }

    fun dot(other: Vector2d): Double {
        return x * other.x + y * other.y
    }

    fun distanceTo(other: Vector2d): Double {
        return this.minus(other).norm()
    }

    fun rotated(angle: Double): Vector2d {
        // 10 * 1              - 0
        val newX = x * Math.cos(angle) - y * Math.sin(angle)
        // 10 * 0              + 0
        val newY = x * Math.sin(angle) + y * Math.cos(angle)
        return Vector2d(newX, newY)
    }

    fun dist(other: Vector2d): Double {
        return sqrt((this.x - other.x).pow(2.0) + (this.y - other.y).pow(2.0))
    }

    override fun toString(): String {
        return java.lang.String.format("(%.3f, %.3f)", x, y)
    }

    fun equals(other: Vector2d): Boolean {
        return if (other is Vector2d) {
            Math.abs(x - other.x) < 1e-4 && Math.abs(y - other.y) < 1e-4
        } else {
            false
        }
    }
}
